### ISSUE TITLE
ceph statefulset

### ISSUE THREAD

**Issue by kfox1111** [rook/rook#2388]
* Bug Report

**Deviation from expected behavior:**
I deployed a test ceph cluster with 0.9 to test it out.
The cluster deployed with Deployment objects.
There is no guarantee that there will never be 2+ of the same pod running at the same time. This could lead to data corruption if not explicitly handled in code. I'm not seeing it though

**Expected behavior:**
Statefulset's should be used instead, to guarantee no more then 1 of each pod is executed at once.

**How to reproduce it (minimal and precise):**
Deploy a cluster. See that important services like each mon member is deployed with a deployment.

**Comment by travisn** [comment]
@kfox1111 The deployments have a replication of 1 and creation strategy of [RecreateDeploymentStrategyType](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#recreate-deployment). We expect that only one instance of the pod will be running in this case.

**Comment by travisn** [comment]
I should have pointed out that in the case there is ever the possibility of running more than one instance of the pod, the mon and osd daemons will fail to start. This can be tested by changing the replication to 2 on the osd and mon deployments and see the pods enter crash loop backoff state.

**Comment by kfox1111** [comment]
the mons and osd's check for this themselves? how do they do that? I could see them doing that with net=host and there being a port conflict. but what about for net!=host?

**Comment by travisn** [comment]
In a quick test of changing the replication to 2 to simulate this issue, I see the following errors when the second daemons trying to start. This is independent from net=host since the errors are from trying to lock the files on the host path.

MON:
```
2018-12-17 15:05:30.082 7fbafe098140  4 rocksdb: [/home/jenkins-build/build/workspace/ceph-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/13.2.2/rpm/el7/BUILD/ceph-13.2.2/src/rocksdb/db/db_impl.cc:252] Shutdown: canceling all background work
2018-12-17 15:05:30.082 7fbafe098140  4 rocksdb: [/home/jenkins-build/build/workspace/ceph-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/13.2.2/rpm/el7/BUILD/ceph-13.2.2/src/rocksdb/db/db_impl.cc:397] Shutdown complete
2018-12-17 15:05:30.083 7fbafe098140 -1 rocksdb: IO error: While lock file: /var/lib/rook/mon-c/data/store.db/LOCK: Resource temporarily unavailable
2018-12-17 15:05:30.083 7fbafe098140 -1 rocksdb: IO error: While lock file: /var/lib/rook/mon-c/data/store.db/LOCK: Resource temporarily unavailable
2018-12-17 15:05:30.083 7fbafe098140 -1 error opening mon data directory at '/var/lib/rook/mon-c/data': (22) Invalid argument
```

OSD:
```
2018-12-17 15:05:49.069 7fda00b9f1c0 -1 osd.0 0 OSD::pre_init: object store '/var/lib/rook/osd0' is currently in use. (Is ceph-osd already running?)
2018-12-17 15:05:49.069 7fda00b9f1c0 -1  ** ERROR: osd pre_init failed: (16) Device or resource busy
```

**Comment by kfox1111** [comment]
ah. ok. That should be safe enough for now then.

There is another issue raised somewhere I saw that was talking about being able to run ceph on top of pv's (useful). In that mode, this may need to be reconsidered for that use case?

**Comment by travisn** [comment]
yes, good question, we will need to ensure this is still safe with the pv's. @galexrt thoughts?

**Comment by Russell-IO** [comment]
This would be handful for manager, when prometheus is enabled, it makes the ceph provided grafana / prometheus graphs go all funny as the pod id has changed, if we used statefulset, then this wouldn't happen

**Comment by kfox1111** [comment]
Still relevant I think.

**Comment by kfox1111** [comment]
Curious if this has been solved now that pv mode is supported? Couldn't this be a problem still?

**Comment by kfox1111** [comment]
Can someone please confirm this is not an issue?

**Comment by satoru-takeuchi** [comment]
@kfox1111 There is a problem in exclusive control between multiple OSD pods. It will be fixed in #6793
